### PR TITLE
Let me Rust that for you

### PR DESCRIPTION
**Pull request by cholcombe973**
I saw your blog post the other day and thought this would be an exciting opportunity to write some Rust code.  This program as far as I can tell produces identical output to NullEscape.c.  

In benchmarking after dropping caches and inserting length 1800 into the content field by following the example in the comments the Rust program is roughly 2x as fast.  I haven't tried to perform any optimization yet.  a.out is the original NullEscape.c with gcc -O3.  null-escape is the Rust program compiled with --release.
```
root@ip-172-31-23-110:~# echo 3 > /proc/sys/vm/drop_caches
root@ip-172-31-23-110:~# echo 2 > /proc/sys/vm/drop_caches
root@ip-172-31-23-110:~# echo 1 > /proc/sys/vm/drop_caches
root@ip-172-31-23-110:~# time cat /tmp/dump/MyTest.txt | ./a.out > test1

real	0m0.193s
user	0m0.003s
sys	0m0.004s
root@ip-172-31-23-110:~# echo 3 > /proc/sys/vm/drop_caches
root@ip-172-31-23-110:~# echo 2 > /proc/sys/vm/drop_caches
root@ip-172-31-23-110:~# echo 1 > /proc/sys/vm/drop_caches
root@ip-172-31-23-110:~# time cat /tmp/dump/MyTest.txt | ./null-escape > test1

real	0m0.068s
user	0m0.000s
sys	0m0.007s
```
I reran these benchmarks several times and got similar results.  I don't know how big of a dataset you're running through this tool so it's possible at much larger values the advantage goes away.  In any event though the Rust program is memory safe among other things :)
If there's interest in using this tool I'd be happy to clean it up a bit so there's no unwraps().  Those can panic.

### CODE REVIEW

**Comment by cholcombe973** [comment]
To build this binary you need to install the stable release of rust: https://www.rust-lang.org/en-US/downloads.html and call `cargo build --release` in the directory.  Cargo will download all the libraries, build the binary and stick in in `null-escape/target/release/null-escape`

**Comment by cholcombe973** [comment]
I added unit tests to prove the functionality is identical.  Rust generics + monomorphism ensures that this won't lose any speed :)

**Comment by rwultsch** [comment]
This is really cool. I will test it this week.

**Comment by jparise** [comment]
Thanks for including a test, @cholcombe973!

File: null-escape/src/main.rs [hunk 1]
```diff
@@ -0,0 +1,173 @@
...
+    assert_eq!(output4, vec![0x5c, 0x5c, 0x5c, 0x5c, 0x30]);
+
+    // 0x5c5c5c3030 => 0x5c5c0030
+    let test5 = &[0x5c, 0x5c, 0x5c, 0x30, 0x30];
+    let mut input5 = std::io::Cursor::new(test5);
+    let mut output5: Vec<u8> = Vec::new();
+    process_input(&mut input5, &mut output5);
+    println!("output5: {:?}", output5);
+    assert_eq!(output5, vec![0x5c, 0x5c, 0x00, 0x30]);
+
```
Comment (jparise) [inline null-escape/src/main.rs]: Stray newline

**Comment by cholcombe973** [comment]
@jparise I've got 2 changes coming.  The first one speeds it up quite a bit.  I found with a bunch of callgrind tests that as the input size grows the program slows down due to memcpy's all over the place.  After that I'm looking into using memchr which should speed it up quite a bit over the baseline.

**Comment by cholcombe973** [comment]
Here's a 275MB random file with the original c program and the rust program compared.  This is big enough that I was seeing a slow down before.  
Original C:
```
time ../a.out < ~/test1 > /dev/null

real	0m2.842s
user	0m2.762s
sys	0m0.076s

```
Rust with better iterator's:
```
time ./target/release/null-escape < ~/test1 > /dev/null

real	0m0.697s
user	0m0.616s
sys	0m0.080s

```

**Comment by ChrisMacNaughton** [comment]
A slightly larger benchmark of the two for you guys:

the testfile:
```bash
dd if=/dev/urandom bs=4M count=250 of=/home/chris/test                                                                                                   130 ↵
250+0 records in
250+0 records out
1048576000 bytes (1.0 GB, 1000 MiB) copied, 79.4674 s, 13.2 MB/s
```
## the results:
Rust:
```bash
time (for i in $(seq 1 10) ; do target/release/null-escape < ~/test > /dev/null; done)
( for i in $(seq 1 10); do; target/release/null-escape < ~/test > /dev/null; )  18.72s user 0.98s system 99% cpu 19.726 total
```
or 1.9727 seconds on average out of 10.

C:
```bash
time (for i in $(seq 1 10) ; do ./a.out < /home/chris/test > /dev/null; done)                                                                            130 ↵
( for i in $(seq 1 10); do; ./a.out < /home/chris/test > /dev/null; done; )  23.94s user 0.95s system 99% cpu 24.908 total
```
or 2.4908 seconds on average out of 10.

Doing a bit more profiling, it looks like the Rust version uses marginally less memory (<1K) than the C version as well.

**Comment by jparise** [comment]
@cholcombe973 That default looks reasonable, but it might be nice to add a command line option that overrides it. That would make it easier to experiment with different values.

File: null-escape/src/main.rs [hunk 2]
```diff
@@ -0,0 +1,218 @@
...
+        Ok(written_bytes)
+    }
+}
+
+fn process_byte<W: Write>(byte: u8, writer: &mut W, count: &mut u64) -> Result<(), std::io::Error> {
+    if byte == 0x30 {
+        if (*count + 1) % 2 == 0 {
+            // we saw 0 or even number of 0x5c before 0x5c30
+            try!(write_byte(0x00, writer));
+            return Ok(());
```
Comment (jparise) [inline null-escape/src/main.rs]: This is my lack of applied rust talking, but why do we `try!` here and below instead of just returning the Result of `write_byte` directly? `process_input` is already matching on `Ok(_)`.

Comment (cholcombe973) [inline null-escape/src/main.rs]: The try! is to catch any IO errors that happen while calling write_byte.  It's used to short circuit execution early on error and push the error back up the stack where we print it in main.  On https://github.com/pinterest/mysql_utils/pull/7/files#diff-432daa1eb52f098d546b14e3a856d4b4R214 that's where the error would get propagated to.  @ChrisMacNaughton any additional thoughts?

Comment (cholcombe973) [inline null-escape/src/main.rs]: You're right though.  This function could return a usize of bytes written and the caller could check to make sure that it is 1 instead of 0.

Comment (jparise) [inline null-escape/src/main.rs]: Right, the root of my thought was noticing that, with the current code factoring, `process_byte` doesn't contain its own loop and is always returning to `process_input`'s byte.

I'm cool with the code either way. I was mainly asking to validate my own understanding. :)

Comment (cholcombe973) [inline null-escape/src/main.rs]: Yeah I think for the try!(write_byte ...) case it's prob ok to ignore the return value because I've said anything less than writing 1 byte is considered an error.  For the other 2 write where we're writing 2 bytes It's prob a good idea to check that 2 bytes were written.

**Comment by jparise** [comment]
I'm excited for this change. Everything looks good to me overall. Let's have @rwultsch give it a try and report back before merging, though.

**Comment by rwultsch** [comment]
tl;dr: I did not compile with -O and therefore C was slow.

So I did a bit of testing based on bdf18ae and the performance of the rust version was terrible. It was roughly 1/8th the speed of the C version. I did some research about how to compile rust and I ended up having done it very wrong. I had just ructc'ed it.

Red is no unescaping
Green is C unescaping
Blue is Rust unescaping.
screen shot 2016-08-25 at 2 33 47 pm

So I did a bit of research and recompiled the rust version the correct way (cargo build --release). Numbers went way up. I then recompiled the C version with -Ofast.

Green is rust.
Red is C.
screen shot 2016-08-25 at 2 35 59 pm
Averaged out, rust is slightly faster than the C version. Both the optimized version are not much slower than running without unescaping.

**Comment by cholcombe973** [comment]
@rwultsch Very cool!  Thanks for testing this out.  Yeah without `cargo build --release` it's going to be crazy slow.  llvm does a ton of optimization on the binary.

**Comment by rwultsch** [comment]
Looking at performance, with the additional flags, i2.4xl servers look IO bound.

**Comment by ChrisMacNaughton** [comment]
@rwultsch that's amazing, on a single NVME SSD on my laptop, I'm CPU bound on this process, with either the C or Rust version, not IO bound

**Comment by ChrisMacNaughton** [comment]
@jparise I did a bit of testing with various buffers between (inclusive) 16k and 128k (on powers of 2) and the 32k performed dramatically better on both my NVME SSD and on AWS EBS drives attached to am m3 medium

**Comment by rwultsch** [comment]
We run CORES/2 processes to dump data. Also, InnoDB data is often not sequential, so we are doing random reads. So we are able to pull in ~18k random 16kb reads per second.

**Comment by rwultsch** [comment]
This pull request ended up being really helpful and made our ETL jobs faster. Based on not much difference at this point from the c to rust, I am going to close this pull request.

**Comment by ChrisMacNaughton** [comment]
These two sentences seem contradictory, you say there's not much difference between the two but that the Rust version ended up being really helpful?

**Comment by rwultsch** [comment]
After recompiling with -Ofast the cpu consumption of the C version dropped like a rock. We are now IO bound rather than CPU.
